src/Rap2hpoutre/LaravelLogViewer/LaravelLogViewer.php

```diff
@@ -173,42 +173,33 @@ public function all()
 
         $file = app('files')->get($this->file);
 
-        preg_match_all($this->pattern->getPattern('logs'), $file, $headings);
+        // Divide log file into an array of logs.
+        preg_match_all($this->pattern->getPattern('logs'), $file, $parts);
 
-        if (!is_array($headings)) {
-            return $log;
-        }
-
-        $log_data = preg_split($this->pattern->getPattern('logs'), $file);
+        foreach($parts[0] ?: [] as $part) {
+            // Divide the log part into a heading and a stack trace.
+            preg_match('/([^\n]*)(.*)/s', $part, $divide);
+            $heading = $divide[1];
+            $stack = $divide[2];
 
-        if ($log_data[0] < 1) {
-            array_shift($log_data);
-        }
+            preg_match($this->pattern->getPattern('heading', ['loglevels' => join('|', $this->level->all())]), $heading, $current);
 
-        foreach ($headings as $h) {
-            for ($i = 0, $j = count($h); $i < $j; $i++) {
-                foreach ($this->level->all() as $level) {
```

Comment:
this old inner loop caused obvious issues since you analyze log entries multiple times. The new solution uses a single main loop which loops each log entry once. For this to work the `$this->pattern->getPattern('heading')` was changed to:
```
^\[(\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}(?:[\+-]\d{4})?)\](?:.*?(\w+)\.|.*?)(debug|info|notice|warning|error|critical|alert|emergency|processed|failed): (.*?)( in .*?:[0-9]+)?$
```
The complexity of this regex is hidden away in 2 variables `$options['date']` and `$options['loglevels']` which are passed to the function.